#55 Auction Settlement

bh2smith opened the pull request:
Closes #55

bh2smith committed:
StateTransition of type Auction

bh2smith committed:
#55 - State Transition of type Auction, with unit tests (full coverage)

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -68,13 +70,39 @@ contract SnappAuction is SnappBase {
+        auctions[slot].appliedAccountStateIndex = stateIndex();
+        
+        emit AuctionSolution(slot, prices, volumes);
+        emit StateTransition(TransitionType.Auction, stateIndex(), _newStateRoot, slot);
```
What's the reasoning behind having two different events here? Isn't this update one atomic transaction that should invoke exactly one callback inside the event listener?

The way it's designed right now, you'd have partial data in two different callbacks.

bh2smith replied:
The reasoning here is to use the existing `StateTransition` event of type Auction. I mostly submitted this to get some feedback and idea about how we might do this differently (i.e. with a single event). It looks like we are going to have to scrap the StateTransition of type Auction...

bh2smith replied:
Done!

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -68,13 +70,39 @@ contract SnappAuction is SnappBase {
+        require(slot == 0 || auctions[slot-1].appliedAccountStateIndex != 0, "Must apply auction slots in order!");
+        require(auctions[slot].shaHash == _orderHash, "Order hash doesn't agree");
+        require(auctions[slot].appliedAccountStateIndex == 0, "Auction already applied");
+        require(block.number > auctions[slot].creationBlock + 20, "Requested order slot is still active");
```
Could the block already be inactive if it was full?

bh2smith replied:
Not according to the definition of active provided here. This imposes a hard limit on the notion of an active slot.

fleupold replied:
Is there a technical reason why we shouldn't allow to process a batch as soon as it's been closed? If I understand correctly a batch is closed as soon as it's full and not necessarily after three minutes.

bh2smith replied:
This is true, we could easily implement this change (although I kind of assumed we would more often reach batch-time before batch capacity). I will add this to the list of todos. Does this imply we are going to do the same for deposits and withdraws?

bh2smith replied:
Done!

bh2smith replied:
Done!

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -68,13 +70,39 @@ contract SnappAuction is SnappBase {
+        bytes32 _newStateRoot,
+        bytes32 _orderHash,
+        uint128[MAX_TOKENS] memory prices,
+        uint128[2*AUCTION_BATCH_SIZE] memory volumes
```
Here I'm wondering if it would make sense to encode the volumes as bytes rather than as a vector like this. We are wasting 7 bytes per order (28bits * 2), which adds up to ~500k gas.

While we haven't done any gas optimisation anywhere, I'm wondering what the advantage of having the volumes and prices decompressed in a readable format like this is.
Otherwise, I'd argue we should just squash all of it into a single blob

bh2smith replied:
Thats cool with me. Shall we use bytes? This will require inclusion of a `BytesLibrary` (similar to that used in our Plasma contract). What better time than now?

josojo replied:
yeah, lets do it

bh2smith replied:
Great! I found that the BytesLib supplier, who wS only publishing to ethpm before, now has an npm version of `BytesLib.sol`. Even though I prefer ethpm, at least we can stay consistent now!

bh2smith replied:
[check it out!](https://github.com/GNSPS/solidity-bytes-utils)

fleupold replied:
Why not use the solidity built in `bytes` type? https://solidity.readthedocs.io/en/v0.5.3/types.html#dynamically-sized-byte-array

bh2smith replied:
Well ya, this is the datatype we are going to use, but if there is any reason to do any byte-manipulation (such as concatenation or slicing), then I would be using the BytesLib to do so. However, I suppose your point is that we don't need to do any bit-manipulation in this context.

bh2smith replied:
We are going to have to assert some kind of length on these bytes (i.e. extra conditions for solution submissions). Is this all cool with everyone?

bh2smith replied:
Turns out, there is no need to validate price and volume sizes since any incorrect submissions can be challenged and fail.

fleupold reviewed (commented):
Nice work! A few suggestions/questions inline.

🖲

josojo reviewed (approved):
Beauty!
with well written tests.

bh2smith commented:
### TODO

- Use bytes for prices and volumes
- Emit only one event on `applyAuction` called `AuctionSettlement` which includes all the information currently emitted in `AuctionSolution` and `StateTransition`
- Remove TransitionType.Auction from other StateTransition possibilities. 

Does this sounds good to you @josojo & @fleupold ?

fleupold commented:
Sounds good to me. Also, let's discuss if we have to prevent batches from being processed for at least 3 minutes, even if they have been completely filled e.g. in the first 30 seconds.

bh2smith commented:
### Updated TODO

- Use bytes for prices and volumes
- Emit only one event on `applyAuction` called `AuctionSettlement` which includes all the information currently emitted in `AuctionSolution` and `StateTransition`
- Remove TransitionType.Auction from other StateTransition possibilities. 
- update definition of "active auction slot" to 
```python
is_active = (block.number < slot.startBlock + 20) AND (slot.size < AUCTION_BATCH_SIZE)
```

bh2smith commented on contracts/SnappAuction.sol:
```diff
@@ -68,13 +70,39 @@ contract SnappAuction is SnappBase {
+        stateRoots.push(_newStateRoot);        
+        auctions[slot].appliedAccountStateIndex = stateIndex();
+        
+        emit AuctionSolution(slot, prices, volumes);
```
Thanks @gnosis-bot! Do you have any suggestion where in the contract this hashed solution should be stored?

bh2smith committed:
apply Auction emits AuctionSettlement event (rather than two different) and now accepts prices and volumes in bytes format.

bh2smith committed:
Update definition of orderSlot.isActive

i.e. isActive <==> block.number <= creationBlock + 20 AND size < CAPACITY

bh2smith committed:
- Hash together prices and volumes and store in shaHash attribute of auction slot

- Reorder validation (because of new shaHash attribute in auction slot)

- update unit tests

bh2smith committed:
Merge branch 'master' into 55_auction_settlement

fleupold reviewed (approved):
Thanks for addressing all the comments. Looks really good!
